design/default-env.md

```diff
@@ -0,0 +1,61 @@
+# RFC: Default Stack Environments
+
+## Background
+
+At the moment, if a `Stack` is defined without `env`, it is said to be an environment-agnostic stack. This technically means that `stack.account` and `stack.region` will resolve to `{ "Ref": "AWS::AccountId" }` and `{ "Ref": "AWS::Region" }`. The implication are that code that relies on parsing one of these values will fail (e.g. the EKS library has an [AMI map](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-eks/lib/ami.ts) keyed by `stack.region`).
+
+Users are able to explicitly "inherit" the CLI's current environment by using a couple of environment variables which are always populated by the CLI:
+
+```ts
+new Stack(this, 'MyStack', { 
+  env: { 
+    account: process.env.CDK_DEFAULT_ACCOUNT, 
+    region: process.env.CDK_DEFAULT_REGION 
+  } 
+}
+```
+
+Prior to introducing explicit support for environment agnostic stacks ([PR](https://github.com/aws/aws-cdk/pull/2922)), falling back to the CLI's current environment was the default behavior if you didn't specify `env`.
+
+The problem with this approach was that if users wanted to use `cdk synth` and use the resulting artifact (template/cloud-assembly) in CI/CD systems, they couldn't rely on the fact that results will be deterministic since they will be dependent on the specific runtime environment. For example, in one environment `AWS_DEFAULT_REGION` can be set to `us-east-1` and in another it will be set to `us-west-2`, which can technically result in a totally different output.
+
+## Problem
+
+The current situation creates undesired friction when getting started with certain modules of the CDK (and potentially 3rd party modules as well that depend on explicit env). For example, if a user wishes to deploy an EKS cluster with `cdk deploy`, they will have to define their stacks with explicit `env` (either specify account/region or use `CDK_DEFAULT_XXX`).
+
+## Opportunity
+
+When using `cdk deploy` directly with an app that has stack(s) without explicit `env`, the synthesized output is an **intermediate artifact** that is never visible. In this case, defaulting to the current account/region is safe and is probably what most users expect.
+
+## Proposal
+
+The proposal is:
+
+1. Add an `--default-env=inherit|explicit|agnostic` switch to the CLI which will control the default env mode.
```

Comment:
Not sure that `inherit` is clear enough. How about `from-creds`?

Reply:
I kind of like `inherit`. It is the terminology often used when it comes to environment variables.